src/engine/Sort.cpp

```diff
@@ -52,30 +55,131 @@ std::string Sort::getDescriptor() const {
+        subTable.numRows(), subTable.numColumns(), deadline_, "Sort operation");
+    return computeResultInMemory(subRes);
+  }
+}
+
+// _____________________________________________________________________________
+Result Sort::computeResultInMemory(std::shared_ptr<const Result> subRes) {
+  runtimeInfo().addDetail("is-external", "false");
 
-  AD_LOG_DEBUG << "Sort result computation..." << endl;
+  // First clone the input table, then sort it in place.
   ad_utility::Timer t{ad_utility::timer::Timer::InitialStatus::Started};
   IdTable idTable = subRes->idTable().clone();
   runtimeInfo().addDetail("time-cloning", t.msecs());
   Engine::sort(idTable, sortColumnIndices_);
 
-  // Don't report missed timeout check because sort is not cancellable
+  // Don't report missed timeout check because the in-memory sort is currently
+  // not cancellable.
+  //
+  // TODO: make the in-memory sort cancellable.
   cancellationHandle_->resetWatchDogState();
   checkCancellation();
 
-  AD_LOG_DEBUG << "Sort result computation done." << endl;
+  // Return the sorted result.
   return {std::move(idTable), resultSortedOn(), subRes->getSharedLocalVocab()};
 }
 
+// _____________________________________________________________________________
+Result Sort::computeResultExternalSort(std::shared_ptr<const Result> subRes,
+                                       bool requestLaziness) {
+  runtimeInfo().addDetail("is-external", "true");
+
+  // Create a unique temporary filename for the external sorter in the index
+  // directory.
+  const std::string& onDiskBase =
+      getExecutionContext()->getIndex().getOnDiskBase();
+  std::ostringstream oss;
+  oss << onDiskBase << ".sort-" << reinterpret_cast<uintptr_t>(this) << "-"
+      << std::this_thread::get_id() << ".dat";
+  std::string tempFilename = oss.str();
+
+  // Get memory limit from the same runtime parameters that we use for the
+  // materialized view writer.
+  //
+  // TODO: Give this runime parameter a more general name.
+  ad_utility::MemorySize memoryLimit =
+      getRuntimeParameter<&RuntimeParameters::materializedViewWriterMemory_>();
+
+  // Get the number of columns from the subtree (this works for both lazy and
+  // materialized results).
+  size_t numColumns = subtree_->getResultWidth();
+
+  // Create the external sorter with our dynamic column comparator. We use a
+  // shared pointer so the sorter survives when the lambda below captures it.
+  using Sorter = ad_utility::CompressedExternalIdTableSorter<SortByColumns, 0>;
+  auto sorter = std::make_shared<Sorter>(
+      tempFilename, numColumns, memoryLimit, allocator(),
+      ad_utility::DEFAULT_BLOCKSIZE_EXTERNAL_ID_TABLE,
+      SortByColumns{sortColumnIndices_});
+
+  // Push all rows into the sorter, handling both lazy and materialized results.
+  // For materialized results, we copy the local vocab, for lazy results we
+  // merge the local vocabs from the individual blocks.
+  auto mergedLocalVocab = std::make_shared<LocalVocab>();
```

Comment:
Is there any benefit to using a shared pointer?

Reply:
When the output is lazy, we capture this in the lambda at the end of the function. Isn't this the way how we usually do this in oder to avoid lifetime issues?

Reply:
You can just move the object into the lambda: `[localVocab = std::move(localVocab)](){ /* code */ }`